### ISSUE TITLE
How to handle named environments with different python versions?

### ISSUE THREAD

**Issue by mhechthz** [astral-sh/uv#9906]
Since I'm new to uv I still did not get how to deal with different named venvs.

I want to have a venv311 and a venv312 in the same project. The python version can be pinned only once and uv sync ignores pre-existing venvs with defined python versions.

How to handle the case of testing different python versions on the same project?

**Comment by nathanscain** [comment]
Curious what the workflow is that is driving this?

uv's cacheing tends to be good enough to flip back and forth without much delay since it is just relinking rather than recopying (especially if you provide `--frozen` to use the existing `uv.lock` resolution).

I know that the `uv venv` command lets you create custom virtual environments to use with `uv pip ...`, but I don't know of a way to tell sync to use a custom environment. Possibly `--python venv311/bin/python`, but I'm not sure if that would have the intended function.

**Comment by zanieb** [comment]
I'd use `uv run --python <version> ...` to test different Python versions. I agree with @nathanscain I'd be curious to understand why you want multiple static environments around.

Regardless, the workaround is 

```
UV_PROJECT_ENVIRONMENT=venv311 uv sync -p 3.11
UV_PROJECT_ENVIRONMENT=venv312 uv sync -p 3.12
```

**Comment by jayqi** [comment]
I have a similar request to what the original poster asked.

My use case in particular is for a project that is a Python library. I want to be able to run tests across all supported Python versions. I recognize that uv manages to make it easy and fast to delete and recreate environments, but this feels kind of weird and is unnecessary disk I/O? 

Supporting multiple virtual environments is something lots of other Python project management tools like [Hatch](https://hatch.pypa.io/latest/config/environment/advanced/#matrix), [Poetry](https://python-poetry.org/docs/managing-environments/), and [PDM](https://pdm-project.org/en/latest/usage/venv/#create-a-virtualenv-yourself) all support. 

What I want to have is a "dev" primary environment as the default one that gets used (this could be the `.venv` one that uv sort of considers the one and only environment right now), and then additional environments for tests for a matrix of Python versions. I believe this is a setup that these other tools have ended up making pretty common.

**Comment by zanieb** [comment]
You can use `--isolated` to avoid changing your default project environment.

**Comment by jayqi** [comment]
> You can use `--isolated` to avoid changing your default project environment.

Awesome! This at least addresses my problem of clobbering the default project environment. Am I understanding correctly that an `--isolated` environment is completely throwaway and not reused? It feels wasteful to keep recreating the test environments, but I suppose uv is fast enough that I can live with this.

**Comment by zanieb** [comment]
Yeah it's thrown away, we could consider changing that but I think we would need a concrete motivating example. It's a goal of uv to make environments disposable via speed.

**Comment by axiomofjoy** [comment]
I develop on a project that supports Python 3.9 through Python 3.13. When CI catches issues affecting a particular Python version, it would be convenient to be able to switch my interpreter in my IDE to a virtual environment corresponding to that version to debug without needed to delete my current environment and reinstall everything. This is something I would have previously done with `conda` by having distinct environments for different Python versions.

**Comment by zanieb** [comment]
Yeah but the same could be said for, e.g., the combination of all dependency groups / extras, which is 2^N environments. I think we need more generalized solutions to these things rather than just creating and storing a bunch of them indefinitely. For example, a uv integration with your IDE could just create the desired environment on request.

**Comment by axiomofjoy** [comment]
> Yeah but the same could be said for, e.g., the combination of all dependency groups / extras, which is 2^N environments. I think we need more generalized solutions to these things rather than just creating and storing a bunch of them indefinitely. For example, a uv integration with your IDE could just create the desired environment on request.

I buy that. If it's fast and doesn't require a lot of manual commands to switch between configurations, I am all for a declarative approach.

**Comment by steve-mavens** [comment]
I have an odd niche case where I have a project on an exFAT-formatted removable device. So, no hardlinks. The performance updating the virtualenv to switch Python versions is of course unusable, so I want to keep environments for each Python version that are either correct, or at least close enough to it that `uv sync` doesn't take 2 minutes. Personally I'd marginally prefer a command-line option over an environment variable, but the workaround is OK. I accept I've done something weird so my punishment is more weirdness ;-)

I suppose I should also try setting UV_PROJECT_ENVIRONMENT to something on the same drive as UV_CACHE_DIR, so that I can leave the environment alone and use `-p` rather than switching it. But I'm new to `uv` and I thought I should at least give "venv in the project directory" a chance...

**Comment by zsol** [comment]
@steve-mavens you might want to take a look at https://github.com/astral-sh/uv/pull/18214 to see if that helps you

**Comment by steve-mavens** [comment]
@zsol: thanks, yes, that might be worth trying too. I think it can probably be argued that a Python virtualenv is a stupid thing for me to keep on an exFAT volume, regardless of the tool chain that got me there!
